### PR TITLE
Reduce duplication in specs using shared_examples

### PR DESCRIPTION
**Pull request by govindkrjoshi**

### CODE REVIEW

File: spec/support/shared_spec_for_sync_controller.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
+require 'rails_helper'
```
Comment (timcheadle) [inline spec/support/shared_spec_for_sync_controller.rb]: I'd move this file to `spec/controllers/shared_examples/` instead.

File: spec/support/shared_spec_for_sync_requests.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
+require 'rails_helper'
```
Comment (timcheadle) [inline spec/support/shared_spec_for_sync_requests.rb]: I'd move this file to `spec/requests/shared_examples/` instead.

File: spec/controllers/api/v1/blood_pressures_controller_spec.rb [hunk 3]
```diff
@@ -1,24 +1,21 @@
...
-        expect(response).to have_http_status(200)
-      end
+  describe 'POST sync: send data from device to server;' do
+    it_behaves_like 'sync controller - create new records'
+    it_behaves_like 'sync controller - update exiting records'
 
+    describe 'creates new blood pressures' do
       it 'creates new blood pressures with associated patient' do
-        patient = FactoryBot.create(:patient)
+        patient         = FactoryBot.create(:patient)
```
Comment (timcheadle) [inline spec/controllers/api/v1/blood_pressures_controller_spec.rb]: This whitespace is unnecessary

File: spec/controllers/api/v1/blood_pressures_controller_spec.rb [hunk 4]
```diff
@@ -1,24 +1,21 @@
...
-    describe 'creates new blood pressures' do
+  let(:build_payload) { lambda { build_blood_pressure_payload } }
+  let(:build_invalid_payload) { lambda { build_invalid_blood_pressure_payload } }
+  let(:update_payload) { lambda { |blood_pressure| updated_blood_pressure_payload blood_pressure } }
 
-      it 'returns 400 when there are no blood pressures in the request' do
-        post(:sync_from_user, params: { blood_pressures: [] })
-        expect(response.status).to eq 400
-      end
+  let(:number_of_schema_errors) { 3 }
```
Comment (timcheadle) [inline spec/controllers/api/v1/blood_pressures_controller_spec.rb]: I don't know what this is for?

File: spec/controllers/api/v1/patients_controller_spec.rb [hunk 5]
```diff
@@ -1,15 +1,18 @@
...
-    describe 'creates new patients' do
+  let(:build_payload) { lambda { build_patient_payload } }
+  let(:build_invalid_payload) { lambda { build_invalid_patient_payload } }
+  let(:update_payload) { lamda { |record| updated_patient_payload record } }
 
-      it 'returns 400 when there are no patients in the request' do
-        post(:sync_from_user, params: { patients: [] })
-        expect(response.status).to eq 400
-      end
+  let(:number_of_schema_errors) { 2 + invalid_record['phone_numbers'].count }
```
Comment (timcheadle) [inline spec/controllers/api/v1/patients_controller_spec.rb]: I definitely don't know what this is for. Where is `invalid_record`?

File: spec/controllers/api/v1/blood_pressures_controller_spec.rb [hunk 6]
```diff
@@ -27,117 +24,10 @@
...
-
-        received_blood_pressures = response_1['blood_pressures'].concat(response_2['blood_pressures']).to_set
-        expect(received_blood_pressures.count).to eq BloodPressure.count
-
-        expect(received_blood_pressures.to_set)
-          .to eq (BloodPressure.all.map {|blood_pressure| Api::V1::Transformer.to_response(blood_pressure) }).to_set
-      end
-    end
+  describe 'GET sync: send data from server to device;' do
+    it_behaves_like 'sync controller - get records'
```
Comment (timcheadle) [inline spec/controllers/api/v1/blood_pressures_controller_spec.rb]: I'd make the shared example title more explicit:

```ruby
it_behaves_like "a working sync controller sending records"
```

File: spec/controllers/api/v1/patients_controller_spec.rb [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
+
+  describe 'POST sync: send data from device to server;' do
+    it_behaves_like 'sync controller - create new records'
```
Comment (timcheadle) [inline spec/controllers/api/v1/patients_controller_spec.rb]: I'd make this example title more explicit as well:

```ruby
it_behaves_like "a working sync controller receiving new records"
```

File: spec/models/address_spec.rb [hunk 8]
```diff
@@ -2,7 +2,6 @@
 
 RSpec.describe Address, type: :model do
   describe 'Validations' do
-    it { should validate_presence_of(:device_created_at)}
-    it { should validate_presence_of(:device_updated_at)}
+    it_behaves_like 'application record'
```
Comment (timcheadle) [inline spec/models/address_spec.rb]: I'd make this explicit as well:

```ruby
it_behaves_like "a record that can be synced remotely"
```

File: spec/controllers/api/v1/blood_pressures_controller_spec.rb (hunk 4 above)
Comment (ssrihari) [reply spec/controllers/api/v1/blood_pressures_controller_spec.rb]: We have renamed this to `number_of_schema_errors_in_invalid_payload`. `invalid_payload` is defined above, and this defines the number of schema errors in that.

File: spec/controllers/api/v1/patients_controller_spec.rb (hunk 5 above)
Comment (ssrihari) [reply spec/controllers/api/v1/patients_controller_spec.rb]: We have extracted `invalid_record` into a `let` above this, and renamed this to `number_of_schema_errors_in_invalid_payload` as well.

**Pull request merged**